### PR TITLE
Add adaptive Muon

### PR DESCRIPTION
**Pull request by dmr51**
Add adaptive Muon with precondition like in AdamW-style. I am Machine Learning Engineer, but not very good in math, so it is needed for a proof that it will work, I wrote this code using ChatGPT. But this variation of adaptive Muon gives 1.5x speedup in convergence in my private benchmark using CIFAR-10 with almost the same accuracy (93.9 vs 94.1).

### CODE REVIEW

**Comment by dmr51** [comment]
I found the paper of AdaMuon https://arxiv.org/abs/2507.11005. It seems close to what I got with ChatGPT, but it does the AdamW step after the orthogonalization (as I understand). As I said, I am not very good in math, but ChatGPT tells that it is better to do AdamW step before the orthogonalization if we do the step elementwise, not layerwise, because if we do it after, it can break the meaning of Muon altogether.

**Comment by parlance-zz** [comment]
>  I wrote this code using ChatGPT. But this variation of adaptive Muon gives 1.5x speedup in convergence in my private benchmark using CIFAR-10 with almost the same accuracy

I've tried a few adaptive scaling methods, including AdaMuon you linked above and what I found is that if you're seeing a big gain it's only because you didn't use a good learn rate schedule to begin with. Your base learn rate is _probably_ too low and your learn rate decay _probably_ isn't aggressive enough.

**Comment by dmr51** [comment]
@parlance-zz What do you think about NorMuon. It looks like even KellerJordan uses it for nanogpt. It does preserve the ortoginalization rowwise, so the meaning of Muon is not broken.

**Comment by dmr51** [comment]
Thanks for the comment about learning rate, btw. It seems to be the case, but I appreciate if you give your opinion about NorMuon.

**Comment by parlance-zz** [comment]
> Thanks for the comment about learning rate, btw. It seems to be the case, but I appreciate if you give your opinion about NorMuon.

No problem! I use NorMuon as well now, the gain over regular Muon is small but tangible. My models use EDM2-style weight normalization, in which matrix params are normalized row-wise as well (the norm is reduced over 1 to ndim), so in my case there might be some extra synergy there. They measure the original orthogonalized update norm and rescale after row-wise normalization so the overall update norm remains the same. It's easy to remove this rescaling to make it more adam-esque, but again, empirically maintaining the original norm and using a better LR schedule beats it every time.

As an aside, I'm not sure why this weight normalization strategy hasn't really caught on, there have been a few popular papers since then that propose essentially the exact same thing, usually with fancier terminology (constraining QKV matrices to the hyper-spherical manifold improves training stability! etc). It also has the added benefit of eliminating weight decay and bias parameters entirely, which is helpful when using Muon, and allows _even higher_ base learn rates without compromising stability.

One last thing I'd mention is if you use grouped convolutions you should consider modifying Muon / NomMuon to be group aware. Separating QKV linear layers into individual matrices improves Muon performance for the exact same reasons. Grouped convolutions are essentially a collection of identical sized smaller submatrices. My implementation of it is here: https://github.com/parlance-zz/dualdiffusion/blob/main/src/training/nor_muon.py

I didn't open a pull request to add it in the official repo(s) because to make it work I'm adding an attribute to the param objects in the model to indicate the number of groups, as pytorch doesn't have a way to keep that information bundled with the parameter (at least when using torch.nn.functional for everything, which I do). Anyway, I didn't think this kind of ad-hoc kludge would be accepted, but the benefits are very real if you do happen to use groups.
